lib/jsonapi_parameters/translator.rb

```diff
@@ -69,11 +51,59 @@ def jsonapi_unsafe_params
     end
   end
 
+  def jsonapi_relationships
+    @jsonapi_relationships ||= @jsonapi_unsafe_hash.dig(:data, :relationships) || []
+  end
+
   def jsonapi_included
     @jsonapi_included ||= @jsonapi_unsafe_hash[:included] || []
   end
 
-  def jsonapi_relationships
-    @jsonapi_relationships ||= @jsonapi_unsafe_hash.dig(:data, :relationships) || []
+  def handle_relationships(param, relationship_key, relationship_value)
+    relationship_value = relationship_value[:data]
+    handler_args = [relationship_key, relationship_value, jsonapi_included]
+    handler = if Handlers.resource_handlers.key?(relationship_key)
+                Handlers.handlers[Handlers.resource_handlers[relationship_key]]
+              else
+                case relationship_value
+                when Array
+                  Handlers.handlers[:to_many]
+                when Hash
+                  Handlers.handlers[:to_one]
+                when nil
+                  Handlers.handlers[:nil]
+                else
+                  raise NotImplementedError.new('relationship resource linkage has to be a type of Array, Hash or nil')
+                end
+              end
+
+    key, val = handler.call(*handler_args)
+
+    param[key] = handle_nested_relationships(val)
+
+    param
+  end
+
+  def handle_nested_relationships(val)
+    # We can only consider Hash relationships (which imply to-one relationship) and Array relationships (which imply to-many).
+    # Each type has a different handling method, though in both cases we end up passing the nested relationship recursively to handle_relationship
+    # (and yes, this may go on indefinitely, basically we're going by the relationship levels, deeper and deeper)
```

Comment:
This part bugs me the most. Technically it may be a vulnerability, as if the user prepares the payload in a way where there's a huge tree of nested relations, it may mean that we are going deep down the stack level. 

However, who are we to judge? Maybe that's good? Most platforms, such as Heroku, will limit the time of execution. There's also "Stack level too deep" error that would pop-up at some point too.

Maybe its a good idea to add some default limiter (like, 10 levels of nested relationship) that can be manually adjusted in the initializer? What do you think?

Reply:
So what would you suggest? Maybe we should not enforce any kind of limiting, but yet provide a way to set a limit?

Reply:
It's a decision between leaving a potential abuse window and creating a possibility that someone will ask us why it doesn't work for more than 2 layers. 
I would pick the latter and:
- set the limit to 2 (so just the top layer and 1 included)
- put an explicit info in readme about it

Reply:
Sounds good to me, although I would extend this to a three level nesting. This seems like a reasonable amount for Rails applications. 

In case there's a user that needs more, what are the options for him? If we do not allow extending it manually (by, for instance, overwriting default LIMIT value), the user will be forced to create a fork (or, in the more negative scenario, use another library).

Reply:
I'm fine with 3 as well.
Definitely it should be possible to change it.